R/options_spec.R

```diff
@@ -72,6 +120,7 @@ option_spec <- function(
       desc = desc,
       option_name = option_name,
       envvar_name = envvar_name,
+      option_fn = option_fn,
```

Comment:
If we pass `...` and `env`, `default`, etc. to `option_fn` in `opt()`, we shall check if the signature of `option_fn` is correct or not. E.g., the assertion functions in the `checkmate` package do not accept ellipses, so the user would need to add a wrapper function or use a proper assertion factory to use checkmate with `options`. If want to help the user, we might use the following helper utility:
```
#' @keywords internal
make_option_fn <- function(option_fn) {
  stopifnot("'option_fn' must be a function" = is.function(option_fn))
  .args <- names(formals(args(option_fn)))
  possible_args <- names(formals(opt))
  if ("..." in .args) {
    list(fn = option_fn, args = possible_args)
  } else {
    list(fn = option_fn, args = c(intersect(.args, possible_args), "..."))
  }
}
```

Reply:
Having the function above, line #123 would be modified to:

```suggestion
      option_fn = make_option_fn(option_fn),
```

Reply:
If we do not want this extra complexity in the `options` package, we can just do a simple assertion in `make_option_fn` and keep the code in `opt()` simple.:
```
#' @keywords internal
make_option_fn <- function(option_fn) {
  stopifnot("'option_fn' must be a function" = is.function(option_fn))
  .args <- names(formals(args(option_fn)))
  if (!"..." in .args) {
    stop("'option_fn' must have '...' in its signature", call. = FALSE)
  }
  option_fn
}
```

Reply:
> E.g., the assertion functions in the `checkmate` package do not accept ellipses, so the user would need to add a wrapper function or use a proper assertion factory to use `checkmate` with `options`.

In such cases, I'd prefer the latter suggestion - where users of `options` just use wrapping functions. 

For situations like yours, where you're using a config file to generate options, I think that this is a perfect use case where your yaml-processing step can be more opinionated and surface a more convenient API that might preprocess some user-provided functions before handing them off to `define_options` using a `make_option_fn` like you've shown here.